File: client/src/engine/checks/regimeComparator.js
/**
 * Check 1: Tax Regime Arbitrage — Old vs New regime comparison.
 *
 * HIGHEST-IMPACT check. Computes fully-optimized old regime tax
 * (with HRA, 80C, 80D, NPS) and compares to new regime tax.
 *
 * Ported from backend/tax_engine/checks/regime_comparator.py
 */

import { FindingStatus, Confidence, createFinding } from '../models.js';
import {
  calculateNewRegimeTax,
  calculateOldRegimeTax,
  computeNewRegimeTaxableIncome,
  computeOldRegimeTaxableIncome,
  calculateHraExemption,
  LIMIT_80C,
  LIMIT_80CCD_1B,
  LIMIT_80D_SELF_BELOW_60,
  LIMIT_80D_SELF_SENIOR,
  LIMIT_80D_PARENTS_BELOW_60,
  LIMIT_80D_PARENTS_SENIOR,
  LIMIT_24B_SELF_OCCUPIED,
} from '../taxUtils.js';

/**
 * Compare old vs new regime with fully optimized deductions.
 * @param {object} salary - SalaryProfile
 * @param {object} [opts]
 * @param {boolean} [opts.parentsSenior=false]
 * @param {boolean} [opts.selfSenior=false]
 * @returns {object} Finding
 */
export function checkRegime(salary, opts = {}) {
  const parentsSenior = opts.parentsSenior || false;
  const selfSenior = opts.selfSenior || false;
  const ageCategory = selfSenior ? 'senior' : 'below_60';
  const fy = salary.financial_year;

  // ── New Regime Tax ──────────────────────────────────────────────────
  const newTaxable = computeNewRegimeTaxableIncome(salary);
  const newTaxResult = calculateNewRegimeTax(newTaxable, fy);
  const newTax = newTaxResult.total_tax;

  // ── Optimized Old Regime Tax ────────────────────────────────────────
  // Calculate optimal HRA exemption
  let optimalHra = 0;
  if (salary.hra_received > 0 && salary.monthly_rent > 0) {
    optimalHra = calculateHraExemption(
      salary.basic_salary,
      salary.hra_received,
      salary.monthly_rent * 12,
      salary.is_metro,
    );
  }

  // Calculate optimal 80C (assume user will fill the gap)
  const optimal80c = LIMIT_80C;

  // Calculate optimal 80D (self + parents insurance)
  // For non-seniors, assume employer group cover handles self — only optimize parents.
  // For seniors (60+), no employer group cover — optimize both self and parents.
  const selfLimit = selfSenior ? LIMIT_80D_SELF_SENIOR : LIMIT_80D_SELF_BELOW_60;
  const parentsLimit = parentsSenior ? LIMIT_80D_PARENTS_SENIOR : LIMIT_80D_PARENTS_BELOW_60;
  const optimal80dTarget = selfSenior ? selfLimit + parentsLimit : parentsLimit;
  const optimal80d = Math.max(salary.deduction_80d, optimal80dTarget);

  // Calculate optimal NPS 80CCD(1B)
  const optimalNps1b = LIMIT_80CCD_1B;

  // Home loan interest (Section 24b) — use actual amount, capped at ₹2L
  const optimal24b = Math.min(salary.deduction_24b, LIMIT_24B_SELF_OCCUPIED);

  // Compute old regime taxable income with optimized deductions
  const oldBreakdown = computeOldRegimeTaxableIncome(salary, {
    hraExemption: optimalHra,
    total80c: optimal80c,
    total80d: optimal80d,
    total80ccd1b: optimalNps1b,
    total24b: optimal24b,
  });
  const oldTaxable = oldBreakdown.taxable_income;
  const oldTaxResult = calculateOldRegimeTax(oldTaxable, fy, ageCategory);
  const oldTax = oldTaxResult.total_tax;

  // ── Compare ─────────────────────────────────────────────────────────
  const savings = newTax - oldTax;
  const recommended = savings > 0 ? 'old' : 'new';

  // Build deductions_needed dict
  const deductionsNeeded = {};
  const current80c = salary.deduction_80c + salary.deduction_80ccc + salary.deduction_80ccd_1;
  if (optimalHra > salary.hra_exemption) {
    deductionsNeeded.hra_exemption = optimalHra;
  }
  const gap80c = LIMIT_80C - current80c;
  if (gap80c > 0) {
    deductionsNeeded.section_80c = optimal80c;
    deductionsNeeded.section_80c_gap = gap80c;
  }
  if (optimal80d > salary.deduction_80d) {
    deductionsNeeded.section_80d = optimal80d;
  }
  if (optimalNps1b > salary.deduction_80ccd_1b) {
    deductionsNeeded.section_80ccd_1b = optimalNps1b;
  }
  if (optimal24b > 0) {
    deductionsNeeded.section_24b = optimal24b;
  }

  if (savings > 0) {
    return createFinding({
      check_id: 'regime_arbitrage',
      check_name: 'Tax Regime Optimization',
      status: FindingStatus.OPPORTUNITY,
      finding: `Switching to old regime with full deductions saves \u20b9${savings.toLocaleString('en-IN')}`,
      savings,
      action: (
        `File ITR under old tax regime for FY ${fy}. ` +
        `Invest in ELSS/PPF for 80C, get parents' health insurance for 80D, ` +
        `and open NPS for 80CCD(1B) before March 31`
      ),
      deadline: 'July 31 (ITR filing) \u2014 but investments needed before March 31',
      confidence: Confidence.DEFINITE,
      explanation: (
        `Your employer applied the new regime (default), resulting in tax of ` +
        `\u20b9${newTax.toLocaleString('en-IN')}. Under the old regime with optimized deductions ` +
        `(HRA \u20b9${optimalHra.toLocaleString('en-IN')} + 80C \u20b9${optimal80c.toLocaleString('en-IN')} + ` +
        `80D \u20b9${optimal80d.toLocaleString('en-IN')} + NPS \u20b9${optimalNps1b.toLocaleString('en-IN')}` +
        (optimal24b > 0 ? ` + Home Loan \u20b9${optimal24b.toLocaleString('en-IN')}` : '') +
        `), your tax drops to \u20b9${oldTax.toLocaleString('en-IN')}.`
      ),
      details: {
        new_regime_tax: newTax,
        new_regime_taxable: newTaxable,
        old_regime_tax: oldTax,
        old_regime_taxable: oldTaxable,
        recommended_regime: recommended,
        old_regime_breakdown: oldBreakdown,
        deductions_needed: deductionsNeeded,
      },
    });
  } else {
    return createFinding({
      check_id: 'regime_arbitrage',
      check_name: 'Tax Regime Optimization',
      status: FindingStatus.OPTIMIZED,
      finding: `New regime is already optimal (saves \u20b9${(-savings).toLocaleString('en-IN')} vs old)`,
      savings: 0,
      action: 'No action needed \u2014 continue with new regime',
      deadline: 'N/A',
      confidence: Confidence.DEFINITE,
      explanation: (
        `New regime tax: \u20b9${newTax.toLocaleString('en-IN')}. ` +
        `Old regime tax (even with optimized deductions): \u20b9${oldTax.toLocaleString('en-IN')}. ` +
        `New regime is better by \u20b9${(-savings).toLocaleString('en-IN')}.`
      ),
      details: {
        new_regime_tax: newTax,
        new_regime_taxable: newTaxable,
        old_regime_tax: oldTax,
        old_regime_taxable: oldTaxable,
        recommended_regime: recommended,
        old_regime_breakdown: oldBreakdown,
      },
    });
  }
}

File: backend/tax_engine/checks/regime_comparator.py
"""Check 1: Tax Regime Arbitrage — Old vs New regime comparison.

This is the HIGHEST-IMPACT check. It computes the fully-optimized old regime
tax (with HRA, 80C, 80D, NPS) and compares it to the new regime tax.

The savings reported here IS the ground-truth regime switch benefit. Individual
checks (80C, 80D, etc.) show component breakdowns for transparency.
"""

from ..models import SalaryProfile, Finding, FindingStatus, Confidence
from ..tax_utils import (
    calculate_new_regime_tax,
    calculate_old_regime_tax,
    compute_new_regime_taxable_income,
    compute_old_regime_taxable_income,
    calculate_hra_exemption,
    get_marginal_rate,
    LIMIT_80C,
    LIMIT_80CCD_1B,
    LIMIT_80D_SELF_BELOW_60,
    LIMIT_80D_SELF_SENIOR,
    LIMIT_80D_PARENTS_BELOW_60,
    LIMIT_80D_PARENTS_SENIOR,
    STANDARD_DEDUCTION,
)


def check_regime(
    salary: SalaryProfile,
    parents_senior: bool = False,
    self_senior: bool = False,
    has_self_health_insurance: bool = False,
) -> Finding:
    """Compare old vs new regime with fully optimized deductions.

    Args:
        salary: Structured salary data from Form 16.
        parents_senior: True if either parent is 60+.
        self_senior: True if taxpayer is 60+.
        has_self_health_insurance: True if user has own health insurance policy.

    Returns:
        Finding with regime recommendation and savings.
    """
    fy = salary.financial_year
    age_category = "senior" if self_senior else "below_60"

    # ── New Regime Tax ──────────────────────────────────────────────────
    new_taxable = compute_new_regime_taxable_income(salary)
    new_tax_result = calculate_new_regime_tax(new_taxable, fy)
    new_tax = new_tax_result["total_tax"]

    # ── Optimized Old Regime Tax ────────────────────────────────────────
    # Calculate optimal HRA exemption
    optimal_hra = 0.0
    if salary.hra_received > 0 and salary.monthly_rent > 0:
        optimal_hra = calculate_hra_exemption(
            basic_annual=salary.basic_salary,
            hra_received_annual=salary.hra_received,
            rent_paid_annual=salary.monthly_rent * 12,
            is_metro=salary.is_metro,
        )

    # Calculate optimal 80C (EPF + fill gap with ELSS/PPF)
    current_80c = salary.deduction_80c + salary.deduction_80ccc + salary.deduction_80ccd_1
    optimal_80c = min(max(current_80c, salary.epf_employee_contribution), LIMIT_80C)
    # If there's still a gap, recommend filling it
    optimal_80c = LIMIT_80C  # Assume user will fill the gap

    # Calculate optimal 80D (self + parents insurance)
    # For non-seniors, assume employer group cover handles self — only optimize parents.
    # For seniors (60+), no employer group cover — optimize both self and parents.
    self_limit = LIMIT_80D_SELF_SENIOR if self_senior else LIMIT_80D_SELF_BELOW_60
    parents_limit = LIMIT_80D_PARENTS_SENIOR if parents_senior else LIMIT_80D_PARENTS_BELOW_60
    optimal_80d_target = (self_limit + parents_limit) if self_senior else parents_limit
    optimal_80d = max(salary.deduction_80d, optimal_80d_target)

    # Calculate optimal NPS 80CCD(1B)
    optimal_nps_1b = LIMIT_80CCD_1B

    # Compute old regime taxable income with optimized deductions
    old_breakdown = compute_old_regime_taxable_income(
        salary,
        hra_exemption=optimal_hra,
        total_80c=optimal_80c,
        total_80d=optimal_80d,
        total_80ccd_1b=optimal_nps_1b,
    )
    old_taxable = old_breakdown["taxable_income"]
    old_tax_result = calculate_old_regime_tax(old_taxable, fy, age_category=age_category)
    old_tax = old_tax_result["total_tax"]

    # ── Compare ─────────────────────────────────────────────────────────
    savings = new_tax - old_tax
    recommended = "old" if savings > 0 else "new"

    # Build deductions_needed dict (only for items not already claimed)
    deductions_needed = {}
    if optimal_hra > salary.hra_exemption:
        deductions_needed["hra_exemption"] = optimal_hra
    gap_80c = LIMIT_80C - current_80c
    if gap_80c > 0:
        deductions_needed["section_80c"] = optimal_80c
        deductions_needed["section_80c_gap"] = gap_80c
    if optimal_80d > salary.deduction_80d:
        deductions_needed["section_80d"] = optimal_80d
    if optimal_nps_1b > salary.deduction_80ccd_1b:
        deductions_needed["section_80ccd_1b"] = optimal_nps_1b

    if savings > 0:
        return Finding(
            check_id="regime_arbitrage",
            check_name="Tax Regime Optimization",
            status=FindingStatus.OPPORTUNITY,
            finding=f"Switching to old regime with full deductions saves \u20b9{savings:,.0f}",
            savings=savings,
            action=(
                f"File ITR under old tax regime for FY {fy}. "
                f"Invest in ELSS/PPF for 80C, get parents' health insurance for 80D, "
                f"and open NPS for 80CCD(1B) before March 31"
            ),
            deadline=f"July 31 (ITR filing) \u2014 but investments needed before March 31",
            confidence=Confidence.DEFINITE,
            explanation=(
                f"Your employer applied the new regime (default), resulting in tax of "
                f"\u20b9{new_tax:,.0f}. Under the old regime with optimized deductions "
                f"(HRA \u20b9{optimal_hra:,.0f} + 80C \u20b9{optimal_80c:,.0f} + "
                f"80D \u20b9{optimal_80d:,.0f} + NPS \u20b9{optimal_nps_1b:,.0f}), "
                f"your tax drops to \u20b9{old_tax:,.0f}."
            ),
            details={
                "new_regime_tax": new_tax,
                "new_regime_taxable": new_taxable,
                "old_regime_tax": old_tax,
                "old_regime_taxable": old_taxable,
                "recommended_regime": recommended,
                "old_regime_breakdown": old_breakdown,
                "deductions_needed": deductions_needed,
            },
        )
    else:
        return Finding(
            check_id="regime_arbitrage",
            check_name="Tax Regime Optimization",
            status=FindingStatus.OPTIMIZED,
            finding=f"New regime is already optimal (saves \u20b9{-savings:,.0f} vs old)",
            savings=0,
            action="No action needed \u2014 continue with new regime",
            deadline="N/A",
            confidence=Confidence.DEFINITE,
            explanation=(
                f"New regime tax: \u20b9{new_tax:,.0f}. "
                f"Old regime tax (even with optimized deductions): \u20b9{old_tax:,.0f}. "
                f"New regime is better by \u20b9{-savings:,.0f}."
            ),
            details={
                "new_regime_tax": new_tax,
                "new_regime_taxable": new_taxable,
                "old_regime_tax": old_tax,
                "old_regime_taxable": old_taxable,
                "recommended_regime": recommended,
                "old_regime_breakdown": old_breakdown,
            },
        )
